File: python/gpr/variables/sources.py
from numpy import exp, inf

from gpr.misc.functions import det3, sigma_norm
from gpr.variables.state import sigma
from gpr.variables.wavespeeds import c_s2


def theta1inv(ρ, A, MP):
    """ Returns 1/θ1
    """
    τ1 = MP.τ1

    if τ1==inf:
        return 0

    cs2 = c_s2(ρ, MP)

    if MP.PLASTIC:
        σY = MP.σY
        n = MP.n
        ρ = det3(A) * MP.ρ0
        σ = sigma(ρ, A, MP)
        sn = sigma_norm(σ)
        if sn == 0:
            return 0
        τ = τ1 * (σY / sn) ** n
    else:
        τ = τ1

    return 3 * det3(A)**(5/3) / (cs2 * τ)

def theta2inv(ρ, T, MP):
    """ Returns 1/θ2
    """
    ρ0 = MP.ρ0
    T0 = MP.T0
    cα2 = MP.cα2
    τ2 = MP.τ2

    return 1 / (cα2 * τ2 * (ρ / ρ0) * (T0 / T))


def K_arr(ρ, λ, T, MP):
    """ Returns the rate of reaction according to Arrhenius kinetics
    """
    Bc = MP.Bc
    Ea = MP.Ea
    Rc = MP.Rc

    return Bc * ρ * λ * exp(-Ea / (Rc*T))

def K_dis(ρ, λ, T, MP):
    """ Returns the rate of reaction according to discrete ignition temperature
        reaction kinetics
    """
    Ti = MP.Ti
    Kc = MP.Kc

    if T > Ti:
        return ρ * λ * Kc
    else:
        return 0

def K_ing(ρ, λ, p, MP):
    """ Returns the rate of reaction according to ignition and growth
        reaction kinetics
    """
    I = MP.I
    G1 = MP.G1
    G2 = MP.G2
    a = MP.a
    b = MP.b
    c = MP.c
    d = MP.d
    e = MP.e
    g = MP.g
    x = MP.x
    y = MP.y
    z = MP.z
    φIG = MP.φIG
    φG1 = MP.φG1
    φG2 = MP.φG2

    φ = 1 - λ

    ret = 0
    if φIG > φ:
        ret += I * λ**b * (ρ-1-a)**x
    if φG1 > φ:
        ret += G1 * λ**c * φ**d * p**y
    if φ > φG2:
        ret += G2 * λ**e * φ**g * p**z

    return ret

File: cpp/src/system/variables/sources.cpp
#include <cmath>

#include "../../etc/globals.h"
#include "../functions/matrices.h"
#include "../functions/vectors.h"
#include "../variables/state.h"
#include "../waves/shear.h"

const int DISCRETE = 0;
const int ARRHENIUS = 1;
const int IGNITION_GROWTH = 2;

double theta1inv(VecVr Q, Par &MP) {
  // Returns the relaxation parameter for the distortion tensor

  // TODO: implement mixing of μ

  double ρ = Q(0);
  Mat3_3Map A = get_A(Q);

  double A53 = pow(A.determinant(), 5. / 3.);
  double cs2 = c_s2(ρ, MP);
  double n = MP.n;

  double sn;
  if (MP.SOLID) {

    if (MP.POWER_LAW) {

      Mat3_3 σ = sigma(Q, MP);
      sn = sigma_norm(σ);
      sn = std::min(sn, 1e8); // Hacky fix

      double σY = MP.σY;
      return 3 * A53 / (cs2 * MP.τ0) * pow((sn / σY), n);
    } else
      return 0.;
  } else if (MP.BINGHAM) {

    double τf = MP.τf;
    double τs = MP.τs;
    double bf2 = MP.bf2;
    double bs2 = MP.bs2;
    double ε = 1e-9;

    Mat3_3 σ = sigma(Q, MP);
    double Γ = sigma_norm(σ) / MP.σY;
    // double Γ = σ.norm() / sqrt(2.) / MP.σY;
    double h = 0.5 * (1. + Γ - sqrt(1. + (4. * ε - 2.) * Γ + Γ * Γ));

    double λ;
    if (Γ < ε)
      λ = 1.;
    else
      λ = h / Γ;
    λ = std::min(λ, 1.);

    double τ = τf * τs / (λ * τf + (1 - λ) * τs);
    double Cs = λ * sqrt(bs2) + (1 - λ) * sqrt(bf2);
    return 3. * A53 / (Cs * Cs * τ);
  } else {
    double c = A53 * MP.ρ0 / (2 * std::pow(MP.μ, 1 / n));

    if (MP.POWER_LAW) {
      Mat3_3 σ = sigma(Q, MP);
      sn = σ.norm() / sqrt(2.);

      return c * pow(sn, (1 - n) / n);
    } else
      return c;
  }
}

double theta2inv(VecVr Q, Par &MP) {
  // Returns the relaxation parameter for the thermal impulse vector

  // TODO: implement mixing of κ

  double ρ = Q(0);
  double T = temperature(Q, MP);
  return T / (MP.κ * ρ);
}

void f_body(Vec3r x, Par &MP) { x = MP.δp; }

double reaction_rate(VecVr Q, Par &MP) {

  double λ = Q(mV) / Q(0);
  λ = std::max(0., λ);

  switch (MP.REACTION) {

  case DISCRETE: {
    if (temperature(Q, MP) > MP.Ti)
      return MP.Kc * λ;
    else
      return 0.;
  }
  case ARRHENIUS: {
    double T = temperature(Q, MP);
    return MP.Bc * exp(-MP.Ea / (MP.Rc * T)) * λ;
  }
  case IGNITION_GROWTH: {
    double p = pressure(Q, MP);
    double tmp1 = std::pow(λ, MP.a);
    double tmp2 = std::pow(2 - λ - MP.λ0, MP.b);
    double tmp3 = std::pow(p, MP.d);
    return MP.G1 * tmp1 * tmp2 * tmp3;
  }
  default:
    throw "Reaction type not recognized";
  }
}
